File: leetcode/239/c++/solution_1.cpp
/*
    239. Sliding Window Maximum
    Author: Miguel Angel Bermeo Ayerbe
*/
class Solution {
   public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        if (k == 1) {
            return nums;
        }
        std::vector<int> output(nums.size() - k + 1, 0);

        for (size_t i = 0, i_max = 0; i < nums.size(); ++i) {
            if (nums[i] >= nums[i_max]) {
                i_max = i;
            }
            if (i >= k - 1) {
                if (i_max < (i + 1) - k) {
                    i_max = (i + 1) - k;
                    for (size_t j = (i + 1) - k + 1; j <= i; ++j) {
                        if (nums[j] >= nums[i_max]) {
                            i_max = j;
                        }
                    }
                }

                output[((i + 1) - k)] = nums[i_max];
            }
        }
        return output;
    }
};

File: leetcode/239/rust/solution_1.rs
/*
    239. Sliding Window Maximum
    Author: Miguel Angel Bermeo Ayerbe
*/
impl Solution {
    pub fn max_sliding_window(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let k_length = k as usize;
        let mut result: Vec<i32> = Vec::with_capacity(nums.len() - k_length + 1 );
        let mut max_i: i32 = 0;
        for i in 0..result.capacity() {
            if i == 0 || (i > 0 && max_i == nums[i-1] && max_i > nums[i+k_length-1]) {
                max_i = *nums[i..(i+k_length)].iter().max().unwrap();
            }else {
                max_i = i32::max(max_i,nums[i+k_length-1] );
            }
            result.push(max_i);
        }
        result
    }
}